### PR TITLE
[CONTINT-4408] add tags to events for all kubernetes resource types

### PR DESCRIPTION
**Pull request by ewoodthomas**
### What does this PR do?

Adds tags to events for all resource types. Specifically, added tags for events that fall under the `deployment://` or `kubernetes_metadata://` entityIDs. Also migrated adding tags for pod events under the `kubernetes_pod_uid://` entityID from `unbundled_events.go` to `events_common.go`.

### Motivation

#27369 added the ability to tag labels and annotations for any kubernetes resource type. These tags should be attached to events involving that resource type.

### Describe how to test/QA your changes

The following QA uses the `container-agent-envs-doc/create_test_env` folder.

- Create a kind cluster
```
1_cluster_management/kind/create.sh
```
- Modify `linux.yaml` in `2_agent_management/helm/values`
```
datadog:
  clusterName: <your-cluster-name>
  ...
agents:
  containers:
    agent:
      env:
        - name: DD_KUBERNETES_RESOURCES_LABELS_AS_TAGS
          value: |
            {"horizontalpodautoscalers.autoscaling":{"hpa-lab":"test-hpa-label"},"deployments.apps":{"x-team":"team-from-label"},"pods":{"x-ref":"reference"},"namespaces":{"kubernetes.io/metadata.name":"name-as-tag"},"nodes":{"kubernetes.io/os":"os-as-tag", "kubernetes.io/arch": "arch-as-tag"}}
        - name: DD_KUBERNETES_RESOURCES_ANNOTATIONS_AS_TAGS
          value: |
            {"horizontalpodautoscalers.autoscaling":{"hpa-ano":"test-hpa-anno"},"deployments.apps":{"x-team":"team-from-annotation"},"pods":{"x-ann":"annotation-reference"}}
  ...
clusterAgent:
  ...
  env:
    - name: DD_KUBERNETES_RESOURCES_LABELS_AS_TAGS
      value: |
        {"horizontalpodautoscalers.autoscaling":{"hpa-lab":"test-hpa-label"},"services":{"service-lab":"test-service-label"},"deployments.apps":{"x-team":"team-from-label"},"pods":{"x-ref":"reference"},"namespaces":{"kubernetes.io/metadata.name":"name-as-tag"},"nodes":{"kubernetes.io/os":"os-as-tag", "kubernetes.io/arch": "arch-as-tag"}}
    - name: DD_KUBERNETES_RESOURCES_ANNOTATIONS_AS_TAGS
      value: |
        {"horizontalpodautoscalers.autoscaling":{"hpa-ano":"test-hpa-anno"},"services":{"service-ano":"test-service-anno"},"deployments.apps":{"x-team":"team-from-annotation"},"pods":{"x-ann":"annotation-reference"}}
  ...
...
```
- This follows the format `<resource_name>.<resource_group>: {<label>:<tag>}`
- Deploy the agent
```
CLUSTER_AGENT_IMAGE_TAG=<cluster-agent-tag> AGENTS_IMAGE_TAG=<agent-tag> 2_agent_management/helm/install.sh
```
- Modify `3_dummy_app_management/k8s/manifests/nginx/deployment-nginx.yaml`
```
...
metadata:
  ...
  labels:
    ...
    x-team: cont-p
  annotations:
    x-team: cont-int
spec:
  ...
  template:
    metadata:
      labels:
        ...
        x-ref: base
      annotations:
        ...
        x-ann: acid
    ...
  ...
...
```
- Modify `3_dummy_app_management/k8s/manifests/nginx/horizontal-pod-autoscaler.yaml`
```
...
apiVersion: autoscaling/v2
kind: HorizontalPodAutoscaler
metadata:
  annotations:
    hpa-ano: dummy-hpa-ano-here
  labels:
    hpa-lab: dummy-hpa-lab-here
  ...
...
```
- Deploy the workload
```
3_dummy_app_management/k8s/deploy.sh
```

- Once everything deploys, go the event explorer and filter by `kube_clutser_name:<your-cluster-name>` and `"nginx"`
- Events with the title `Events from the Pod...` should have the tags `annotation-reference:acid` and `reference:base`
- Events with the title `Events from the Deployment...` should have the tags `team-from-label:cont-p` and `team-from-annotation:cont-int`
- Events with the title `Events from the HorizontalPodAutoscaler...` should have the tags `test-hpa-anno:dummy-hpa-anno-here` and `test-hpa-label:dummy-hpa-lab-here`
- All these events should have tags `name-as-tag:<namespace>`, `os-as-tag:<os>`, and `arch-as-tag<arch>`. However, the last two sometimes take a while to start showing up on events.

### CODE REVIEW

File: pkg/collector/corechecks/cluster/kubernetesapiserver/events_common.go
Comment (adel121) [inline pkg/collector/corechecks/cluster/kubernetesapiserver/events_common.go:274]: Not urgent at all, but just as a headsup:

Thanks to [this](https://github.com/DataDog/datadog-agent/pull/28645) PR, we will be able to create the entity ids with the NewEntityID function exposed by the tagger, which can avoid typos in the id's and also have a performance impact.

You will be able to create the entity id as follows:
`types.NewEntityID(types.KubernetesPodUID, involvedObjects.UID)` 

[Here](https://github.com/DataDog/datadog-agent/blob/ba27a220b174de70d63d38302067f1c9b70f3358/pkg/process/util/containers/containers.go#L122) is an example.

This applies to all the lines below where you create the id as a string with `fmt.Sprintf`

I'd suggest one of 2 options:
- Not waiting for the other PR to be merged, and merging your PR as it is, but create another card (in CONINT board or CONTP board) to initialise the entity id here with the NewEntityID function. We can do it later when the other PR is merged.
- Don't merge your PR for now, and wait some time until the other PR is merged, then modify your code and merge it. The other PR should be merged very soon, I think tomorrow in the worst case. 

If you have any questions, feel free to ping me.

File: pkg/collector/corechecks/cluster/kubernetesapiserver/events_common.go
Comment (adel121) [inline pkg/collector/corechecks/cluster/kubernetesapiserver/events_common.go:286]: I think you added this line for debugging. Please don't forget to remove it.,

File: pkg/collector/corechecks/cluster/kubernetesapiserver/events_common.go
Comment (adel121) [inline pkg/collector/corechecks/cluster/kubernetesapiserver/events_common.go:285]: Is there a way to get the plural name directly? 

appending an `"s"` can be risky, some resources in kubernetes won't respect this rule, especially CRDs.

Comment (ewoodthomas) [reply pkg/collector/corechecks/cluster/kubernetesapiserver/events_common.go:285]: The involved object doesn't directly expose the resource type. Looked into how we could directly get the resource type, and the only option I've found is to request for all the resources from the Discovery API and sort through them until finding the resource with the matching `Kind` field. We do a similar thing [here](https://github.com/DataDog/datadog-agent/blob/0318c751827c0513dfb1e5f4786076de35f58d5d/comp/core/workloadmeta/collectors/internal/kubeapiserver/utils.go#L156), but I'm guessing we don't want to be doing this every time we go to get the tags. Seeing if we could potentially create a map of `kind --> resource` when the collector starts.

Comment (adel121) [reply pkg/collector/corechecks/cluster/kubernetesapiserver/events_common.go:285]: Discussed on zoom and agreed to implement a component that facilitates introspection in kubernetes resources, it can use caching, and can be shared by other components. 

For now, we can proceed by appending an "s" and we will create a separate task to work on implementing the shared introspection solution.

File: pkg/collector/corechecks/cluster/kubernetesapiserver/events_common.go
Comment (sblumenthal) [inline pkg/collector/corechecks/cluster/kubernetesapiserver/events_common.go:286]: If we need the EntityID as a string, could we use the `GenerateKubeMetadataEntityID` [function instead](https://github.com/DataDog/datadog-agent/blob/023a2a4ae55b6a42dac1ed0cdf6ddab3be606619/comp/core/workloadmeta/collectors/util/kube_metadata_util.go#L15-L20)?

**Review by sblumenthal — APPROVED** [review]
One small comment, overall the change looks good to me!

**Pull request merged**
